### DISCUSSION TITLE
React Hot reloading

### DISCUSSION THREAD

**Discussion by pepicrft** [ElMassimo/vite_ruby#58] [Q&A] [answered]
Hi there!
First of all, thanks for building such a great Gem. I was able to move away from Webpack easily and I'm now using Vite with a React frontend. This is the entrypoint of my app:

```jsx
import React from "react";
import ReactDom from "react-dom";
import App from "../components/App";

ReactDom.render(<App />, document.getElementById("react-app"));
```

The only thing that it does is mounting a React component in the DOM. 

Although everything works fine, I noticed a couple of things that I was wondering whether they can be improved through a plugin or a different configuration on my side:
- When I modify React components, the frontend doesn't reload automatically.
- If I make a small change in one of the components and reload, the bundling through Vite takes a long time.

Am I doing something wrong?

**Comment by pepicrft** [comment]
In case it helps debug the issue, [here's the PR](https://github.com/craftweg/buildify/pull/4) where I'm adding Vite.

**Comment by ElMassimo** [comment]
Hi Pedro!

This is something I should add to the docs. As stated in the [Vite docs](https://vitejs.dev/guide/backend-integration.html#backend-integration), the react-refresh plugin currently relies on `HTML` modifications that [can't be performed when using server-rendered HTML](https://github.com/vitejs/vite-plugin-react/pull/11#discussion_r430879201).

If you check the console in your example app, you should probably see something like:

<img width="1439" alt="Screen Shot 2021-04-21 at 12 14 13" src="https://user-images.githubusercontent.com/1158253/115577955-381d3180-a29b-11eb-84a2-df82eac7259e.png">

What this means, is that you'll need to add the following in development:

```html
<script type="module">
    import RefreshRuntime from "http://localhost:3000/@react-refresh"
    RefreshRuntime.injectIntoGlobalHook(window)
    window.$RefreshReg$ = () => {}
    window.$RefreshSig$ = () => (type) => type
    window.__vite_plugin_react_preamble_installed__ = true
</script>
```

**Comment by ElMassimo** [comment, marked as answer]
I have added a `vite_react_refresh_tag` helper that you can use to [set up HMR with React](https://vite-ruby.netlify.app/guide/troubleshooting.html#hot-module-refresh-does-not-work-for-react).

The helper should be used in any layouts or views that use React components:

```erb
  <%= vite_client_tag %>
  <%= vite_react_refresh_tag %>
  <%= vite_javascript_tag 'application' %>
```

If you locked the gems optimistically using `~>`, you can upgrade by running `bundle update vite_ruby vite_rails`.


https://user-images.githubusercontent.com/1158253/115637268-1eeda280-a2e6-11eb-88a6-360c14ebccc9.mp4
